blueprints/snapshots.and.monitoring.md

```diff
@@ -0,0 +1,828 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Snapshots and Monitoring Separation
+
+## Problem Description
+Traffic Monitor uses data partially from CDN Snapshots and partially from its
+own Monitoring payloads to determine what to monitor and how. This leads to
+race conditions and certain parts of it being out-of-sync with others as one
+data set is polled while another grows stale.
+
+## Proposed Change
+Traffic Router should receive Snapshots directly from Traffic Ops, while
```

Comment:
> Traffic Router should receive Snapshots directly from Traffic Ops

The TO->TM->TR flow is very intentional, to make sure the Monitor is monitoring what the Router is serving.

Removing that flow path and making independent TO->TM and TO->TR flows will create more even worse race conditions than the one mentioned above.

What does TR do if it gets CRConfig changes before the Monitor? So health data doesn't exist for a cache? What if something important, like a Profile, is changed? So the thing being monitored, the server, thresholds, deliveryservices, anything, is different than what the Router needs to serve? Any and all data--routing, thresholding, services, anything--is now a potential race condition, and can lead to strange and very-difficult-to-debug errors, including outages. Worse, if either the Monitor or Router is having trouble getting the latest CRConfig for any reason, those inscrutable errors could be indefinite.

> Traffic Monitor uses data partially from CDN Snapshots and partially from its own Monitoring payloads to determine what to monitor and how. This leads to race conditions

This can be solved either put all Monitoring data in the CRConfig, or to add a "bulk fetch" method to TO to get both objects in one request. 

If necessary, the CRConfig Monitoring data TR doesn't need could be stripped before sending it to TR. But I suspect it's a tiny size increase, and not nearly worth the safety and correctness benefit of having the Monitor directly serve the raw CRConfig bytes it got from TO. It also gives us more flexibility in the future to make TR use that data somehow.

Reply:
Looks like you missed the conversation we had about this in Slack. This was @rawlinp's response to more or less exactly what you just said:

>>  If we are getting CrConfig from TO and CrStates from TM, it is possible that a cache that has been set ADMIN_DOWN could still be health in CrStates.

> that race doesn't seem like a big issue, because once TM has processed the ADMIN_DOWN-ing of the cache, it will set it unhealthy in CrStates, but in the meantime TR will still be sending traffic to it. This is no different than TR waiting to get the latest CRConfig+health directly from TM -- it will still send traffic to the cache until TM has done its thing.

>>  It's also possible that a cache that has been moved to OFFLINE and should not be in the CrConfig is still in CrStates.

> This isn't really an issue either, because TR only looks up health state for caches that are in its crconfig. I was just in that code a month ago to fix the bug where if a cache went from admin_down to offline, TM removed it from the crstates, causing TR to treat it as available.On the flipside, for caches going from offline to reported/admin_down, if TR has the cache in its crconfig but not in crstates, TR should treat it as unavailable until it gets a valid state.

which I think is at least as well as I could've put it. Originally the plan was actually to have TM generate configuration for TR, but this convinced me and @dneuman64 that it was okay to let TR go directly to TO.

Frankly I don't really care one way or the other, though as @limited has pointed out the latter has some advantages that would be nice to take advantage of.

Reply:
The Server.Status is not the only potential race. As I said

> What if something important, like a Profile, is changed? So the thing being monitored, the server, thresholds, deliveryservices, anything, is different than what the Router needs to serve? Any and all data--routing, thresholding, services, anything--is now a potential race condition

We've discussed adding smarter health, a percentage instead of a boolean, better thresholding, and many other features. Every new feature like those is another potential outage-inducing race, if TM and TR are racing for the same data instead of passing it through a flow.

Reply:
The only data both TM and TR need is which caches to monitor/route to. If those get out of sync, TR has contingencies in place to not route traffic to caches it has no health information about. Basically any cache it knows about via the CRConfig that _doesn't have health info yet_ is considered unavailable. And if TR gets health information about a cache it doesn't know about (because it's not in its CRConfig), it ignores it. So although TM/TR get out of sync with each other until they've both processed the config, it's not a problem.

TM should only be telling TR which caches are healthy or not. I don't see how changing profiles, thresholds, etc., is a potential outage-inducing race condition. Those would only affect how TM determines cache availability, which is all that TR really cares about from TM.

Reply:
The Monitor also has Stat data, and does Delivery Service Thresholding, as well as serving stats to be aggregated via Traffic Stats to an upstream location, which can potentially be critical that it be correct. For example, if it were used for Billing.

What happens if that Delivery Service stat and thresholding is updated on the Router before the Monitor? Now we're routing to things the Monitor doesn't know about, and doesn't know to aggregate, and thus that traffic data is lost upstream, with potential financial or legal ramifications.

> TM should only be telling TR which caches are healthy or not. I don't see how changing profiles, thresholds, etc., is a potential outage-inducing race condition. Those would only affect how TM determines cache availability, which is all that TR really cares about from TM.

TM also serves Delivery Service availability to TR, as above. Moreover, as I said, we'd like to be able to add more complex health, such as DS-specific availability for each cache. These races make that either impossible, or worse, someone might not realize their subtlety and implement constantly racing and error-prone things.

Reply:
TR wouldn't be getting any kind of stat or thresholding data from TO or TM. TM processes that data then tells TR simply "this cache is unavailable" or "this delivery service is unavailable". If we wanted to enhance that to have TM say "this cache is 50% available" for instance, then that still wouldn't be a race w/ TR. That's just TR waiting to get health data from TM as always.